### ISSUE TITLE
AMDVLK driver hangs and crashes upon running vulkan applications on R5 M430

### ISSUE THREAD

**Issue by darhal** [GPUOpen-Drivers/AMDVLK#199]
I tried to build and run  Sascha Willems Vulkan samples after installing the official amdgpu driver from AMD's website, the Vulkan SDK from LunarG. However the driver crashes and hangs the whole PC after rending few frames.

Here is the system informations
```omar@dell:~$ inxi -F
System:    Host: dell Kernel: 5.4.0-54-generic x86_64 bits: 64 Desktop: Gnome 3.36.4 Distro: Ubuntu 20.04.2 LTS (Focal Fossa) 
Machine:   Type: Laptop System: Dell product: Inspiron 15-3567 v: N/A serial: <superuser/root required> 
           Mobo: Dell model: 0JTTXY v: A00 serial: <superuser/root required> UEFI: Dell v: 2.13.0 date: 08/13/2020 
Battery:   ID-1: BAT0 charge: 33.7 Wh condition: 33.7/41.4 Wh (81%) 
CPU:       Topology: Dual Core model: Intel Core i3-6006U bits: 64 type: MT MCP L2 cache: 3072 KiB 
           Speed: 2000 MHz min/max: 400/2000 MHz Core speeds (MHz): 1: 2000 2: 2000 3: 2000 4: 2000 
Graphics:  Device-1: Intel Skylake GT2 [HD Graphics 520] driver: i915 v: kernel 
           Device-2: AMD Sun XT [Radeon HD 8670A/8670M/8690M / R5 M330 / M430 / Radeon 520 Mobile] driver: amdgpu 
           v: 5.6.20.20.45 
           Display: x11 server: X.Org 1.20.9 driver: amdgpu resolution: 1920x1080~60Hz 
           OpenGL: renderer: Mesa DRI Intel HD Graphics 520 (SKL GT2) v: 4.6 Mesa 20.2.6 
Audio:     Device-1: Intel Sunrise Point-LP HD Audio driver: snd_hda_intel 
           Sound Server: ALSA v: k5.4.0-54-generic 
Network:   Device-1: Qualcomm Atheros QCA9565 / AR9565 Wireless Network Adapter driver: ath9k 
           IF: wlp2s0 state: up mac: d8:0f:99:6a:63:b7 
           Device-2: Realtek RTL810xE PCI Express Fast Ethernet driver: r8169 
           IF: enp3s0 state: down mac: 10:7d:1a:01:36:b2 
           Device-3: Qualcomm Atheros type: USB driver: btusb 
Drives:    Local Storage: total: 945.92 GiB used: 24.58 GiB (2.6%) 
           ID-1: /dev/sda vendor: Seagate model: ST1000LM035-1RK172 size: 931.51 GiB 
           ID-2: /dev/sdb type: USB vendor: Kingston model: DataTraveler 3.0 size: 14.41 GiB 
Partition: ID-1: / size: 452.26 GiB used: 22.94 GiB (5.1%) fs: ext4 dev: /dev/sda3 
Sensors:   System Temperatures: cpu: 47.0 C mobo: 47.0 C gpu: amdgpu temp: 40 C 
           Fan Speeds (RPM): cpu: 0 
Info:      Processes: 264 Uptime: 23m Memory: 3.71 GiB used: 1.95 GiB (52.7%) Shell: bash inxi: 3.0.38 
```

Here is the relevant kernal logs from `/var/log/kern.log`. The logs shows everything related to amdgpu driver from booting to the execution of the vulkan app (@`330.244727`) the driver crashes and hangs the whole PC after rendering few frames (I enabled `gpu_recovery` to avoid the hanging).
```Feb  5 19:15:24 dell kernel: [  539.346911] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  5 19:16:57 dell kernel: [    0.000000] Linux version 5.4.0-54-generic (buildd@lcy01-amd64-024) (gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)) #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 (Ubuntu 5.4.0-54.60-generic 5.4.65)
Feb  5 19:16:57 dell kernel: [   13.102524] amdkcl: loading out-of-tree module taints kernel.
Feb  5 19:16:57 dell kernel: [   13.102578] amdkcl: module verification failed: signature and/or required key missing - tainting kernel
Feb  5 19:16:57 dell kernel: [   15.670577] [drm] amdgpu kernel modesetting enabled.
Feb  5 19:16:57 dell kernel: [   15.670578] [drm] amdgpu version: 5.6.20.20.45
Feb  5 19:16:57 dell kernel: [   15.670601] amdgpu: vga_switcheroo: detected switching method \_SB_.PCI0.GFX0.ATPX handle
Feb  5 19:16:57 dell kernel: [   15.671018] amdgpu: CRAT table not found
Feb  5 19:16:57 dell kernel: [   15.671021] amdgpu: Virtual CRAT table created for CPU
Feb  5 19:16:57 dell kernel: [   15.671039] amdgpu: Topology: Add CPU node
Feb  5 19:16:57 dell kernel: [   15.695873] amdgpu 0000:01:00.0: remove_conflicting_pci_framebuffers: bar 0: 0xc0000000 -> 0xcfffffff
Feb  5 19:16:57 dell kernel: [   15.695875] amdgpu 0000:01:00.0: remove_conflicting_pci_framebuffers: bar 2: 0xd0000000 -> 0xd003ffff
Feb  5 19:16:57 dell kernel: [   15.695898] amdgpu 0000:01:00.0: enabling device (0000 -> 0003)
Feb  5 19:16:57 dell kernel: [   15.696023] amdgpu 0000:01:00.0: amdgpu: Trusted Memory Zone (TMZ) feature not supported
Feb  5 19:16:57 dell kernel: [   15.696024] amdgpu 0000:01:00.0: amdgpu: set kernel compute queue number to 8 due to invalid parameter provided by user
Feb  5 19:16:57 dell kernel: [   15.696837] kfd kfd: amdgpu: HAINAN  not supported in kfd
Feb  5 19:16:57 dell kernel: [   15.747220] amdgpu 0000:01:00.0: amdgpu: Fetched VBIOS from ATRM
Feb  5 19:16:57 dell kernel: [   15.747223] amdgpu: ATOM BIOS: BR46915.001
Feb  5 19:16:57 dell kernel: [   15.876935] amdgpu 0000:01:00.0: amdgpu: VRAM: 2048M 0x000000F400000000 - 0x000000F47FFFFFFF (2048M used)
Feb  5 19:16:57 dell kernel: [   15.876937] amdgpu 0000:01:00.0: amdgpu: GART: 256M 0x000000FF00000000 - 0x000000FF0FFFFFFF
Feb  5 19:16:57 dell kernel: [   15.877131] [drm] amdgpu: 2048M of VRAM memory ready
Feb  5 19:16:57 dell kernel: [   15.877136] [drm] amdgpu: 3804M of GTT memory ready.
Feb  5 19:16:57 dell kernel: [   15.878015] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  5 19:16:57 dell kernel: [   15.936727] [drm] amdgpu: dpm initialized
Feb  5 19:16:57 dell kernel: [   16.149390] amdgpu 0000:01:00.0: amdgpu: SE 1, SH per SE 1, CU per SH 5, active_cu_number 5
Feb  5 19:16:57 dell kernel: [   16.542055] [drm] Initialized amdgpu 3.40.0 20150101 for 0000:01:00.0 on minor 1
Feb  5 19:17:13 dell kernel: [   38.682416] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  5 19:21:52 dell kernel: [  318.377008] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  5 19:22:04 dell kernel: [  330.244727] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx timeout, signaled seq=63, emitted seq=65
Feb  5 19:22:04 dell kernel: [  330.244968] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process Renderer pid 7133 thread Renderer pid 7133
Feb  5 19:22:04 dell kernel: [  330.244977] amdgpu 0000:01:00.0: amdgpu: GPU reset begin!
Feb  5 19:22:05 dell kernel: [  330.809225] amdgpu 0000:01:00.0: amdgpu: PCI CONFIG reset
Feb  5 19:22:05 dell kernel: [  330.809228] amdgpu 0000:01:00.0: amdgpu: GPU pci config reset
Feb  5 19:22:05 dell kernel: [  330.813005] amdgpu 0000:01:00.0: amdgpu: GPU reset succeeded, trying to resume
Feb  5 19:22:05 dell kernel: [  330.814439] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  5 19:22:05 dell kernel: [  331.419852] amdgpu 0000:01:00.0: amdgpu: recover vram bo from shadow start
Feb  5 19:22:05 dell kernel: [  331.420569] amdgpu 0000:01:00.0: amdgpu: recover vram bo from shadow done
Feb  5 19:22:05 dell kernel: [  331.420616] amdgpu 0000:01:00.0: amdgpu: GPU reset(1) succeeded!
Feb  5 19:22:16 dell kernel: [  341.509669] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx timeout, signaled seq=67, emitted seq=69
Feb  5 19:22:16 dell kernel: [  341.510062] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process Renderer pid 7133 thread Renderer pid 7133
```
Here is another crash log (from gears in VulkanSamples):
```
Feb  7 14:53:50 dell kernel: [  898.151356] [drm] PCIE gen 3 link speeds already enabled
Feb  7 14:53:50 dell kernel: [  898.152680] amdgpu 0000:01:00.0: amdgpu: PCIE GART of 256M enabled (table at 0x000000F400000000).
Feb  7 14:54:03 dell kernel: [  910.371464] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx timeout, signaled seq=32, emitted seq=34
Feb  7 14:54:03 dell kernel: [  910.371583] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process gears pid 12858 thread gears pid 12858
Feb  7 14:54:03 dell kernel: [  910.371588] amdgpu 0000:01:00.0: amdgpu: GPU reset begin!
Feb  7 14:54:03 dell kernel: [  910.883321] Asynchronous wait on fence i915:gnome-shell[3482]:629c timed out (hint:intel_atomic_commit_ready+0x0/0x54 [i915])
Feb  7 14:54:03 dell kernel: [  910.883376] Asynchronous wait on fence drm_sched:gfx:16 timed out (hint:submit_notify+0x0/0x80 [i915])
```

Vulkan ICD file:
```
omar@dell:~$ cat /etc/vulkan/icd.d/amd_icd64.json
{
  "file_format_version": "1.0.0",
  "ICD": {
        "library_path" : "/opt/amdgpu/lib/x86_64-linux-gnu/amdvlk64.so",
    "api_version": "1.2.152"
  }
}
```

Firmware ME version:
```omar@dell:~$ sudo cat /sys/kernel/debug/dri/1/amdgpu_firmware_info
VCE feature version: 0, firmware version: 0x00000000
UVD feature version: 0, firmware version: 0x00000000
MC feature version: 0, firmware version: 0x00a07730
ME feature version: 29, firmware version: 0x00000091
PFP feature version: 29, firmware version: 0x00000054
CE feature version: 29, firmware version: 0x0000003d
RLC feature version: 1, firmware version: 0x00000002
RLC SRLC feature version: 0, firmware version: 0x00000000
RLC SRLG feature version: 0, firmware version: 0x00000000
RLC SRLS feature version: 0, firmware version: 0x00000000
MEC feature version: 0, firmware version: 0x00000000
SOS feature version: 0, firmware version: 0x00000000
ASD feature version: 0, firmware version: 0x00000000
TA XGMI feature version: 0, firmware version: 0x00000000
TA RAS feature version: 0, firmware version: 0x00000000
SMC feature version: 0, firmware version: 0x15040000
SDMA0 feature version: 0, firmware version: 0x00000000
SDMA1 feature version: 0, firmware version: 0x00000000
VCN feature version: 0, firmware version: 0x00000000
DMCU feature version: 0, firmware version: 0x00000000
DMCUB feature version: 0, firmware version: 0x00000000
VBIOS version: BR46915.001
```

**Comment by Darhal** [comment]
Its probably also important to add that even with gpu_recovery enabled (after the driver crashes due to a VK app running), on shutting down the PC, it hangs in a black screen that show these lines and I have to turn off the power to shut the computer down:
```
kernel: [ 1169.446678] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx timeout, signaled seq=124, emitted seq=126
kernel: [ 1169.446832] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process  pid 0 thread  pid 0
```

**Comment by Flakebi** [comment]
Hi,
I think your GPU is too old to support Vulkan, at least I didn’t find any note that says it is supported. Searching for 8670M on the [driver download site](https://www.amd.com/en/support/graphics/amd-radeon-hd/amd-radeon-hd-8000m-series/amd-radeon-hd-8670m-series-gpu) lists OpenGL 4.4 and DirectX 11 as the latest supported APIs (and the latest driver is from end of 2015).
For completeness, if I go the version numbers right, Sun XT corresponds to gfx6.

That said, did you try if radv works? (The Vulkan driver that is part of mesa.) It should be registered at `/etc/vulkan/icd.d/radeon_icd64.json`.

**Comment by Darhal** [comment]
Hello, thanks for the reply.

Actually my GPU is "AMD Radeon(TM) R5 M430 2GB DDR 3" (this is according to the laptop vendor). In the BIOS it shows R5 M430 as well. The following website shows my GPU as well: https://vulkan.gpuinfo.org/ so it does indeed support Vulkan.

I have tried radv driver, it shows a warning then crashes as well upon creating the device.

The 'amdvlk' driver actually renders few frames before crashing.

**Comment by Flakebi** [comment]
Aha, thanks for the clarification. From what I heard internally, gfx6 is unfortunately not officially supported by amdvlk (the Linux driver, Windows seems to be supported). It is GCN 1.0, so it *could* work, but we’re not testing it.

If you want to debug that further, [umr](https://gitlab.freedesktop.org/tomstdenis/umr) is a good starting point. If you disable gpu recovery and ssh into the machine, it can dump some information.
`umr -t` gives an overview over which parts of the hardware are active (the ones that display 100 % are hanging).
`umr -O bits,disasm_anyways --waves gfx_0.0.0` shows the currently running shader and info like if the shader segfaulted.

**Comment by Darhal** [comment]
`sudo umr -t` yields the following:
```
ERROR: Device 0x1916 not found in UMR device table
ASIC not found (instance=0, did=ffffffffffffffff)
```

**Comment by Flakebi** [comment]
Hm, that makes debugging more difficult.
Does it hang for every Vulkan application? Can you try if `vulkaninfo` and `vkcube` are working? They should be part of the Vulkan SDK or a vulkan-tools package if there is one.

I’m not sure how to grasp the issue, my current ideas are
- if you can find a previous driver version that works, it can be bisected to a change that introduced the regression (though that assumes it is a regression)
- you can try to pinpoint it to some specific app behavior that introduce the hang, the [Conformance Test Suite](https://github.com/KhronosGroup/VK-GL-CTS/blob/master/external/vulkancts/README.md) is a good start as it tests many different features in isolation

**Comment by Darhal** [comment]
Hello, sorry for the late response.

Sascha Willems examples and my engine makes the GPU hangs. I tried to run steam using vulkan, it crashed reporting that it can't find swapchain extensions)

I managed to get the [vkvia.zip](https://github.com/GPUOpen-Drivers/AMDVLK/files/6141935/vkvia.zip) and the [vulkaninfo.txt](https://github.com/GPUOpen-Drivers/AMDVLK/files/6141932/vulkaninfo.txt) files.
I have tried to run ```vkcube --gpu_number 1``` and it worked perfectly (0 being intel's iGPU and 1 the AMD one).

I still have to run the CTS, I will let you know when I get the results.
